x-pack/plugins/ingest_pipelines/public/application/pipeline_processors_editor/pipeline_processors_editor.tsx

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { i18n } from '@kbn/i18n';
+import React, {
+  FunctionComponent,
+  useState,
+  useMemo,
+  useRef,
+  useEffect,
+  MutableRefObject,
+} from 'react';
+import {
+  EuiPanel,
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiDragDropContext,
+  EuiDroppable,
+  EuiDraggable,
+  EuiIcon,
+  EuiTitle,
+  EuiSpacer,
+  EuiButton,
+  EuiButtonEmpty,
+} from '@elastic/eui';
+
+import { Processor } from '../../../common/types';
+
+import { FormFlyout } from './components';
+import { prepareDataIn } from './data_in';
+import { prepareDataOut, DataOutResult } from './data_out';
+import { useEditorState } from './reducer';
+import { PipelineEditorProcessor } from './types';
+
+export interface Props {
+  processors: Processor[];
+  stateReaderRef: (state: MutableRefObject<() => DataOutResult>) => void;
+  onFailure?: Processor[];
+}
+
+export const PipelineProcessorsEditor: FunctionComponent<Props> = ({
+  processors: originalProcessors,
+  stateReaderRef,
+}) => {
+  const dataInResult = useMemo(() => prepareDataIn({ processors: originalProcessors }), [
+    originalProcessors,
+  ]);
+  const [state, dispatch] = useEditorState(dataInResult);
+  const { processors } = state;
+  const [selectedProcessor, setSelectedProcessor] = useState<PipelineEditorProcessor | undefined>(
+    undefined
+  );
+  const [isAddingNewProcessor, setIsAddingNewProcessor] = useState<boolean>(false);
+  const stateGetterRef = useRef(() => prepareDataOut(state));
+
+  useEffect(() => {
+    stateGetterRef.current = () => prepareDataOut(state);
+  }, [state]);
+
+  useEffect(() => {
+    stateReaderRef(stateGetterRef);
```

Comment:
I have a hard time understanding the code here. It would be simpler IMO to call it `onStateGetterChange`, and call it only when it changes (when the `state` changes). WDYT?

```js
useEffect(() => {
  stateGetterRef.current = () => prepareDataOut(state);
  onStateGetterChange(stateGetterRef);
}, [state]);
```

Reply:
This will introduce another dependency to the `useEffect` which I was trying to avoid (because then we have to make sure we do not create an infinite update loop).

The idea here was to more closely follow the pattern of `<input ref={inputRef => /* use the ref */ }`. This frees the component using the pipeline editor from needing to handle state updates emitted from the child component.

If we wanted a mechanism like that I would rather have a prop called `onStateChange` or something like that.

Reply:
Yes, then the following is even much easier to follow

```js
useEffect(() => {
  onStateGetterChange(() => prepareDataOut(state));
}, [state]);
```

It is not a `onStateChange` because it receives a getter function. It's the getter function that changes.

I am not sure I follow the need of having a ref.

Reply:
And now that I look closer, in the consumer component, _there_ we would need a ref and not a state (the component does not need to re-render when that value changes).

```js
// this
const processorStateReaderRef = useRef<any>();

// instead of
const [processorStateReaderRef, setProcessorStateReaderRef] = useState<any>();

// And used like this
<PipelineProcessorsEditor
  onStateGetterChange={stateGetter => processorStateReaderRef.current = stateGetter}
  processors={processors}
/>
```

Reply:
> And now that I look closer, in the consumer component, there we would need...

Hm, I see what you're saying. I think either way we are doing something a little bit unconventional to get data out, the only issue I would still like to avoid is this:

```ts
useEffect(() => {
  onStateGetterChange(() => prepareDataOut(state));
}, [state, /* missing dep */]);
```

While a consumer of pipeline editor may always choose to use a ref like we do here I think it would still be a better API if we follow the pattern of `<el ref={myRef => ... }/>`. My latest commit should contain my take on what I think will meet both our goals best.